app/domain/authentication/authn_gcp/decoded_token.rb

```diff
@@ -17,31 +17,60 @@ def initialize(decoded_token_hash:, logger:)
         @decoded_token_hash = decoded_token_hash
         @logger = logger
 
-        @project_id = token_claim_value(PROJECT_ID_TOKEN_CLAIM_NAME)
-        @instance_name = token_claim_value(INSTANCE_NAME_TOKEN_CLAIM_NAME)
-        @service_account_id = token_claim_value(SUB_TOKEN_CLAIM_NAME)
-        @service_account_email = token_claim_value(EMAIL_TOKEN_CLAIM_NAME)
-        @audience = token_claim_value(AUDIENCE_TOKEN_CLAIM_NAME)
+        initialize_required_claims
+        initialize_optional_claims
       end
 
       private
 
-      def token_claim_value(token_claim)
-        token_claim_path = token_claim.split('/')
-        token_claim_value = @decoded_token_hash.dig(*token_claim_path)
+      def initialize_required_claims
+        @audience = required_token_claim_value(AUDIENCE_TOKEN_CLAIM_NAME)
+        @service_account_id = required_token_claim_value(SUB_TOKEN_CLAIM_NAME)
+        @service_account_email = required_token_claim_value(EMAIL_TOKEN_CLAIM_NAME)
+      end
+
+      def initialize_optional_claims
+        @project_id = optional_token_claim_value(PROJECT_ID_TOKEN_CLAIM_NAME)
+        @instance_name = optional_token_claim_value(INSTANCE_NAME_TOKEN_CLAIM_NAME)
+      end
+
+      def required_token_claim_value(required_token_claim)
```

Comment:
Move closer to caller

Reply:
I personally prefer it that way, if we will get more comment about this i will change it